zcash_primitives/src/transaction/mod.rs

```diff
@@ -0,0 +1,179 @@
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use sapling_crypto::redjubjub::Signature;
+use std::io::{self, Read, Write};
+
+use serialize::Vector;
+
+mod components;
+
+#[cfg(test)]
+mod tests;
+
+use self::components::{Amount, JSDescription, OutputDescription, SpendDescription, TxIn, TxOut};
+
+const OVERWINTER_VERSION_GROUP_ID: u32 = 0x03C48270;
+const OVERWINTER_TX_VERSION: u32 = 3;
+const SAPLING_VERSION_GROUP_ID: u32 = 0x892F2085;
+const SAPLING_TX_VERSION: u32 = 4;
+
+/// A Zcash transaction.
+pub struct Transaction {
+    overwintered: bool,
+    version: u32,
+    version_group_id: u32,
+    vin: Vec<TxIn>,
+    vout: Vec<TxOut>,
+    lock_time: u32,
+    expiry_height: u32,
+    value_balance: Amount,
+    shielded_spends: Vec<SpendDescription>,
+    shielded_outputs: Vec<OutputDescription>,
+    joinsplits: Vec<JSDescription>,
+    joinsplit_pubkey: [u8; 32],
+    joinsplit_sig: [u8; 64],
```

Comment:
These fields are set to zeros when not present in the transaction, whereas the `binding_sig` is encoded as an `Option`. Why the difference?

Reply:
It's a bit of an inconsistency between the available set of primitives here vs in `zcashd`. In `zcashd` we use zeroed arrays in the struct for all three cases, and don't fill them if we have nothing to fill with, same as here. But in Rust, we have an actual `Signature` struct for `binding_sig`, and so it makes more sense there to have an `Option<Signature>` instead of relying on the internals of `Signature` representing the "no signature" case. We should also probably use `Option<[u8; 32]>` etc. on these fields as well, to be clearer about their absence as-necessary (and make checking logic easier to write).

Reply:
Pushed a commit to make these `Option<[u8; N]>`.